subsys/event_manager/event_manager_priv.h

```diff
@@ -201,19 +218,109 @@ extern "C" {
 #define _EVENT_TYPE_DEFINE_SIZES(ename)
 #endif
 
-#define _EVENT_TYPE_DEFINE(ename, init_log_en, log_fn, trace_data_pointer)		\
-	_EVENT_SUBSCRIBERS_ARRAY_TAGS(ename);						\
-	STRUCT_SECTION_ITERABLE(event_type, _CONCAT(__event_type_, ename)) = {		\
-		.name            = STRINGIFY(ename),					\
-		.subs_start      = _EVENT_SUBSCRIBERS_START_TAG(ename),			\
-		.subs_stop       = _EVENT_SUBSCRIBERS_END_TAG(ename),			\
-		.init_log_enable = init_log_en,						\
-		.log_event       = (IS_ENABLED(CONFIG_LOG) ? (log_fn) : (NULL)),	\
-		.trace_data      = (IS_ENABLED(CONFIG_EVENT_MANAGER_PROFILER_TRACER) ?	\
-					(trace_data_pointer) : (NULL)),			\
-		_EVENT_TYPE_DEFINE_SIZES(ename) /* No comma here intentionally */	\
+/** @brief Event header.
+ *
+ * When defining an event structure, the event header
+ * must be placed as the first field.
+ */
+struct event_header {
+	/** Linked list node used to chain events. */
+	sys_snode_t node;
+
+	/** Pointer to the event type object. */
+	const struct event_type *type_id;
+};
+
+/** Function to log data from this event. */
+typedef void (*log_event_data)(const struct event_header *eh);
+
+/** Deprecated function to log data from this event. */
+typedef	int (*log_event_data_dep)(const struct event_header *eh, char *buf, size_t buf_len);
+
+
+
+#if IS_ENABLED(CONFIG_EVENT_MANAGER_USE_DEPRECATED_LOG_FUN)
+#define _EVENT_TYPE_DEFINE_LOG_FUN(log_fun)						\
+	.log_event_func_dep = ((IS_ENABLED(CONFIG_LOG) && Z_C_GENERIC) ?		\
+					((log_event_data_dep)_Generic((log_fn),		\
+						log_event_data : NULL,			\
+						log_event_data_dep : log_fn,		\
+						default : NULL))			\
+					: (NULL)),					\
+	.log_event_func     = (log_event_data) (IS_ENABLED(CONFIG_LOG) ?		\
+					(Z_C_GENERIC ?					\
+						(_Generic((log_fn),			\
+							log_event_data : log_fn,	\
+							log_event_data_dep : NULL,	\
+							default : NULL))		\
+						: (log_fn))				\
+					: (NULL)),
+#else
+#define _EVENT_TYPE_DEFINE_LOG_FUN(log_fun)						\
```

Comment:
Consider emitting build warning if user tries to use deprecated logging function, but the CONFIG_EVENT_MANAGER_USE_DEPRECATED_LOG_FUN=n. With current implementation the log would be silently dropped which could be quite confusing.